core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -286,19 +286,6 @@ protected void doRun() throws Exception {
                 });
             }
         }
-
-        // delete local indices that do neither exist in previous cluster state nor part of tombstones
-        for (AllocatedIndex<? extends Shard> indexService : indicesService) {
-            Index index = indexService.index();
-            IndexMetaData indexMetaData = event.state().metaData().index(index);
-            if (indexMetaData == null) {
-                assert false : "index" + index + " exists locally, doesn't have a metadata but is not part"
-                    + " of the delete index list. \nprevious state: " + event.previousState().prettyPrint()
-                    + "\n current state:\n" + event.state().prettyPrint();
-                logger.warn("[{}] isn't part of metadata but is part of in memory structures. removing", index);
-                indicesService.deleteIndex(index, "isn't part of metadata (explicit check)");
-            }
-        }
```

Comment:
The `state.blocks().disableStatePersistence()` in `IndicesClusterStateService.clusterChanged` removes the indices and then returns early, not processing anything else.  I am not sure that is what we want in the case of having a new cluster, but having missed the earlier cluster state updates, we are now processing this cluster state update with the new cluster UUID. It seems to me that we still want to process the cluster state update through the various methods (create indices, delete indices, etc). 

I do agree that we lose the assertion regarding this case in specific of a new cluster.  I'm not sure that this alone is worth keeping the code for. We could keep the loop and the assertion, and let the removal happen in `removeUnallocatedIndices`?  Or add a comment to `removeUnallocatedIndices` as well as an assertion such as:

```
private void removeUnallocatedIndices(final ClusterChangeEvent event) {
        // same as before

        for (AllocatedIndex<? extends Shard> indexService : indicesService) {
            Index index = indexService.index();
            if (indicesWithShards.contains(index) == false) {
                // if the index does not exist in the cluster state, then it should have been deleted earlier, unless its a brand new cluster (new cluster UUID)
                assert event.state.metaData().index(index) != null || event.isNewCluster(); 
                logger.debug("{} removing index, no shards allocated", index);
                indicesService.removeIndex(index, "removing index (no shards allocated)");
            }
        }
    }